MirrorPeer spec changes

vbnrh opened the pull request:
This commit changes the scheduling interval to a list of intervals.

The common package has been renamed to utils and all the files has been renamed to represent their function.

GowthamShanmugam commented on controllers/utils/s3.go:
```diff
@@ -10,6 +10,22 @@ import (
 	BucketGenerateName = "odrbucket"
+
+	// rook
+	RookOrigin = "rook"
```
RookOrigin is not exactly  S3. 
  Ideally, we should create OriginMap instead of keeping separate variables for each Origin: RookOrigin, S3Origin.

vbnrh replied:
removed the consts and created a map for the origin

umangachapagain commented on api/v1alpha1/mirrorpeer_types.go:
```diff
@@ -52,19 +52,22 @@ type MirrorPeerSpec struct {
 	// +kubebuilder:validation:Optional
-	// +kubebuilder:default="5m"
-	SchedulingInterval string `json:"schedulingInterval,omitempty"`
+	SchedulingInterval []string `json:"schedulingInterval,omitempty"`
```
This won't work. All upgrades will be broken.

GowthamShanmugam replied:
it's better we deprecate this and create a new variable. if a new variable is empty then take value from SchedulingInterval as default.

umangachapagain reviewed (changes requested):
Could you please add more details about why each of these changes are required?
Also, break changes into separate sensible commits and add tests to ensure we don't regress and break upgrades.

vbnrh commented:
/retest

vbnrh commented:
> Could you please add more details about why each of these changes are required? Also, break changes into separate sensible commits and add tests to ensure we don't regress and break upgrades.

1st Commit
Majority of the changes are changing the package name common to utils and its usages throughout the code.
This is done to follow common k8s conventions and have clean naming system.
```
├── hash.go
├── predicate.go
├── s3.go
└── secret.go
```
Also many of the files were name with hyphen ( secret-utils, common-controller-utils ) which violate the go file naming conventions. The file names have been changed to follow proper snake_case now.


2nd Commit
This changes includes the spec changes for MirrorPeer and part of the changes for DR.
Also updates the unit test files associated. So far does not break anything in tests but breaks operator upgrades.

Subject to further discussion and change.

3rd Commit 
Generated files

vbnrh commented on addons/token-exchange/agent_mirrorpeer_controller.go:
```diff
@@ -220,47 +221,52 @@ func (r *MirrorPeerReconciler) enableCSIAddons(ctx context.Context, namespace st
+			return err
+		}
+		op, err := controllerutil.CreateOrUpdate(ctx, r.SpokeClient, found, func() error {
+			found.Spec = vrc.Spec
```
done

vbnrh commented on addons/token-exchange/agent_mirrorpeer_controller.go:
```diff
@@ -220,47 +221,52 @@ func (r *MirrorPeerReconciler) enableCSIAddons(ctx context.Context, namespace st
+				Provisioner: fmt.Sprintf(ProvisionerTemplate, scr.Namespace),
+				Parameters:  params,
+			},
+		}
```
done

umangachapagain commented:
> > Could you please add more details about why each of these changes are required? Also, break changes into separate sensible commits and add tests to ensure we don't regress and break upgrades.
> 
> 1st Commit Majority of the changes are changing the package name common to utils and its usages throughout the code. This is done to follow common k8s conventions and have clean naming system.
> 
> ```
> ├── hash.go
> ├── predicate.go
> ├── s3.go
> └── secret.go
> ```
> 
> Also many of the files were name with hyphen ( secret-utils, common-controller-utils ) which violate the go file naming conventions. The file names have been changed to follow proper snake_case now.
> 
> Also moved the constants and functions that were spread around to their respective util file.
> 
This commit should be a separate PR with this explanation. Also, let's ensure we all agree on the refactors so that we don't end up renaming packages every release.

vbnrh commented:
> > > Could you please add more details about why each of these changes are required? Also, break changes into separate sensible commits and add tests to ensure we don't regress and break upgrades.
> > 
> > 
> > 1st Commit Majority of the changes are changing the package name common to utils and its usages throughout the code. This is done to follow common k8s conventions and have clean naming system.
> > ```
> > ├── hash.go
> > ├── predicate.go
> > ├── s3.go
> > └── secret.go
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Also many of the files were name with hyphen ( secret-utils, common-controller-utils ) which violate the go file naming conventions. The file names have been changed to follow proper snake_case now.
> > Also moved the constants and functions that were spread around to their respective util file.
> 
> This commit should be a separate PR with this explanation. Also, let's ensure we all agree on the refactors so that we don't end up renaming packages every release.

I have removed the refactoring till we resolve the bug. Will send it in next PR.

umangachapagain commented on addons/token-exchange/agent_mirrorpeer_controller.go:
```diff
@@ -225,42 +227,42 @@ func (r *MirrorPeerReconciler) createVolumeReplicationClass(ctx context.Context,
+		params[SchedulingIntervalKey] = interval
+		params[ReplicationSecretNameKey] = ReplicationSecretName
+		params[ReplicationSecretNamespaceKey] = scr.Namespace
+		vrcName := fmt.Sprintf(VolumeReplicationClassNameTemplate, common.FnvHash(interval))
```
why do we need to hash this? for same interval value, it will always collide. Instead let's directly append interval?

vbnrh replied:
@Madhu-1 If two intervals are same, do they point towards same VRC or is it a one to one mapping ?

vbnrh replied:
@umangachapagain I prefer using  hashed value for naming the VRC instead of appending the interval. This will avoid the users from creating their own VRCs and abstract  our logic

umangachapagain replied:
Users can create their own VRC.
No harm in keeping it I guess.

Madhu-1 replied:
its not useful to have 2 VR with same intervals.

umangachapagain commented on addons/token-exchange/agent_mirrorpeer_controller.go:
```diff
@@ -225,42 +227,42 @@ func (r *MirrorPeerReconciler) createVolumeReplicationClass(ctx context.Context,
+			Name: found.Name,
+		}, found)
+
+		if err != nil && !errors.IsNotFound(err) {
+			return err
+		}
```
return all errors after you have gone through all the intervals. don't block other VRC due to 1 failure.

vbnrh replied:
done

umangachapagain commented on addons/token-exchange/agent_mirrorpeer_controller.go:
```diff
@@ -225,42 +227,42 @@ func (r *MirrorPeerReconciler) createVolumeReplicationClass(ctx context.Context,
-			Name: "odf-rbd-volumereplicationclass",
-		},
+		}
+		op, err := controllerutil.CreateOrUpdate(ctx, r.SpokeClient, found, func() error {
+			found.Spec = spec
+			return nil
+		})
```
We should not update a VRC. @Madhu-1 WDYT?

Madhu-1 replied:
Yes only create it no need to update it for now.

vbnrh replied:
Changed the code to create and forget

umangachapagain commented on config/crd/bases/multicluster.odf.openshift.io_mirrorpeers.yaml:
```diff
@@ -67,19 +67,6 @@ spec:
-                type: string
               schedulingInterval:
                 default: 5m
                 description: The SchedulingInterval in which the mirroring snapshot
```
these should be updated as well.

vbnrh replied:
updated description

umangachapagain commented on addons/token-exchange/agent_mirrorpeer_controller.go:
```diff
@@ -50,14 +49,17 @@ type MirrorPeerReconciler struct {
+	ReplicationSecretNamespaceKey      = "replication.storage.openshift.io/replication-secret-namespace"
+	ProvisionerTemplate                = "%s.rbd.csi.ceph.com"
+	RookConfigMapName                  = "rook-ceph-operator-config"
+	VolumeReplicationClassNameTemplate = "odf-rbd-volumereplicationclass-%v"
```
```suggestion
	VolumeReplicationClassNameTemplate = "rbd-volumereplicationclass-%v"
```

vbnrh replied:
ack

umangachapagain commented on addons/token-exchange/agent_mirrorpeer_controller_test.go:
```diff
@@ -37,9 +38,7 @@ var (
-			SchedulingInterval:    "10m",
-			ReplicationSecretName: "my-super-secret-credentials",
-			Mode:                  "journal",
+			SchedulingInterval: []string{"10m", "5m", "30m", "1h"},
```
add tests when list contains same values, or no values.

vbnrh replied:
I have added duplicate values in this but i don't see keeping an empty list as it won't enter the loop in anycase.

umangachapagain commented on api/v1alpha1/mirrorpeer_types.go:
```diff
@@ -52,19 +50,11 @@ type MirrorPeerSpec struct {
 	// +kubebuilder:validation:Optional
-	// +kubebuilder:default="5m"
-	SchedulingInterval string `json:"schedulingInterval,omitempty"`
+	SchedulingInterval []string `json:"schedulingInterval,omitempty"`
```
```suggestion
	SchedulingIntervals []string `json:"schedulingInterval,omitempty"`
```

umangachapagain replied:
CRD doesn't reflect this change. Update it.

vbnrh replied:
updated

Madhu-1 commented on addons/token-exchange/agent_mirrorpeer_controller.go:
```diff
@@ -50,14 +49,17 @@ type MirrorPeerReconciler struct {
+	SchedulingIntervalKey              = "schedulingInterval"
+	ReplicationSecretNameKey           = "replication.storage.openshift.io/replication-secret-name"
+	ReplicationSecretNamespaceKey      = "replication.storage.openshift.io/replication-secret-namespace"
+	ProvisionerTemplate                = "%s.rbd.csi.ceph.com"
```
`ProvisionerTemplate` to `RBDProvisionerName`

vbnrh replied:
done

Madhu-1 commented on addons/token-exchange/agent_mirrorpeer_controller.go:
```diff
@@ -50,14 +49,17 @@ type MirrorPeerReconciler struct {
+	ProvisionerTemplate                = "%s.rbd.csi.ceph.com"
+	RookConfigMapName                  = "rook-ceph-operator-config"
+	VolumeReplicationClassNameTemplate = "odf-rbd-volumereplicationclass-%v"
+	ReplicationSecretName              = "rook-csi-rbd-provisioner"
```
RBDReplicationSecretName.

Madhu-1 replied:
will this name remain the same for external mode also?

vbnrh replied:
Yes currently we are keeping it the same. We are not supporting external mode for now.

Madhu-1 commented on controllers/common/hash.go:
```diff
@@ -0,0 +1,18 @@
+	https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function
+*/
+func FnvHash(s string) uint64 {
+	h := fnv.New64a()
```
use 32 to keep the length minimal? and add a test case for this one

vbnrh replied:
Makes sense to keep the length to minimum but there will be more room for collisions. I am changing it to 32 for time being

Madhu-1 commented on addons/token-exchange/agent_mirrorpeer_controller_test.go:
```diff
@@ -2,12 +2,13 @@ package addons
 import (
 	"context"
+	"fmt"
+	"github.com/red-hat-storage/odf-multicluster-orchestrator/controllers/common"
```
move this to line 11 for import groupings

vbnrh replied:
done

Madhu-1 commented on addons/token-exchange/agent_mirrorpeer_controller.go:
```diff
@@ -225,42 +227,42 @@ func (r *MirrorPeerReconciler) createVolumeReplicationClass(ctx context.Context,
+		})
+		if err != nil {
+			klog.Error(err, "Failed to create/update volume replication class")
+			return err
```
as you are returning here you dont need else block just log it as sucessful

vbnrh replied:
changed

Madhu-1 commented on addons/token-exchange/agent_mirrorpeer_controller.go:
```diff
@@ -225,42 +227,42 @@ func (r *MirrorPeerReconciler) createVolumeReplicationClass(ctx context.Context,
-			Name: "odf-rbd-volumereplicationclass",
-		},
+		}
+		op, err := controllerutil.CreateOrUpdate(ctx, r.SpokeClient, found, func() error {
```
update to specific parameters might not be supported in the future see https://github.com/csi-addons/volume-replication-operator/issues/88

umangachapagain replied:
If we do create only, how do we handle upgrades where VRC spec might change?
We might need to label VRCs so that we can get list of VRC that we created and update it.

Madhu-1 replied:
You need to delete and recreate. VRC is the same as storageclass

vbnrh replied:
Good info to know. We will just create and forget then.

Madhu-1 commented on addons/token-exchange/agent_mirrorpeer_controller.go:
```diff
@@ -219,49 +220,57 @@ func (r *MirrorPeerReconciler) enableCSIAddons(ctx context.Context, namespace st
+		}
+		err = r.SpokeClient.Create(ctx, &vrc)
+		if err != nil {
+			klog.Error(err, "Failed to create VolumeReplicationClass: %s", vrcName)
+			errs = append(errs, err)
+		}
+		klog.Infof("VolumeReplicationClass created: %s", vrcName)
```
this will log both failure and success as there is no break or return statement.

vbnrh replied:
fixed with continue

Madhu-1 commented on controllers/common/hash.go:
```diff
@@ -0,0 +1,18 @@
+	fnv64a is a 64-bit non-cryptographic hash algorithm with a low collision and a high distribution rate.
+	https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function
+*/
+func FnvHash(s string) uint32 {
```
unit test for this one?

vbnrh replied:
added unit test for this

Madhu-1 commented on addons/token-exchange/agent_mirrorpeer_controller.go:
```diff
@@ -21,7 +21,6 @@ import (
 	"crypto/sha1"
 	"encoding/hex"
 	"fmt"
-
```
follow import order https://github.com/golang/go/wiki/CodeReviewComments#imports

vbnrh replied:
somehow goland always messes this up. 
I just installed goimports. This should be fixed now

Madhu-1 commented on addons/token-exchange/agent_mirrorpeer_controller.go:
```diff
@@ -50,14 +49,17 @@ type MirrorPeerReconciler struct {
+	ReplicationSecretNamespaceKey      = "replication.storage.openshift.io/replication-secret-namespace"
+	RBDProvisionerTemplate             = "%s.rbd.csi.ceph.com"
+	RookConfigMapName                  = "rook-ceph-operator-config"
+	VolumeReplicationClassNameTemplate = "rbd-volumereplicationclass-%v"
+	ReplicationSecretName              = "rook-csi-rbd-provisioner"
```
```suggestion
        RBDVolumeReplicationClassNameTemplate = "rbd-volumereplicationclass-%v"
        RBDReplicationSecretName              = "rook-csi-rbd-provisioner"
```

vbnrh replied:
changed

Madhu-1 commented on addons/token-exchange/agent_mirrorpeer_controller.go:
```diff
@@ -50,14 +49,17 @@ type MirrorPeerReconciler struct {
+	SchedulingIntervalKey              = "schedulingInterval"
+	ReplicationSecretNameKey           = "replication.storage.openshift.io/replication-secret-name"
+	ReplicationSecretNamespaceKey      = "replication.storage.openshift.io/replication-secret-namespace"
+	RBDProvisionerTemplate             = "%s.rbd.csi.ceph.com"
```
where `RBDProvisionerTemplate` is set when creating the volume replication class?

vbnrh replied:
Good catch ! Got lost in the refactorings.
Added a testcase for this !

Madhu-1 commented on addons/token-exchange/agent_mirrorpeer_controller_test.go:
```diff
@@ -2,9 +2,10 @@ package addons
 
 import (
 	"context"
+	"fmt"
```
follow import order as mentioned earlier

Madhu-1 reviewed (commented):
left small suggestions.

Madhu-1 reviewed (approved):
LGTM

GowthamShanmugam commented:
LGTM

umangachapagain commented on api/v1alpha1/mirrorpeer_types.go:
```diff
@@ -52,19 +50,11 @@ type MirrorPeerSpec struct {
 	// +kubebuilder:validation:Optional
-	// +kubebuilder:default="5m"
-	SchedulingInterval string `json:"schedulingInterval,omitempty"`
+	SchedulingIntervals []string `json:"schedulingInterval,omitempty"`
```
```suggestion
	SchedulingIntervals []string `json:"schedulingIntervals,omitempty"`
```

umangachapagain commented on bundle/manifests/multicluster.odf.openshift.io_mirrorpeers.yaml:
```diff
@@ -67,24 +67,12 @@ spec:
-                description: The ReplicationSecretName contains the ceph user credentials
-                  to connect to ceph cluster for mirroring operations
-                type: string
               schedulingInterval:
```
still not updated.

vbnrh replied:
Updated

umangachapagain commented:
/cherry-pick release-4.10
